src/2015-04-25-epidemiological-compartment-model.cpp

```diff
@@ -0,0 +1,240 @@
+ * modelers - the sorts of folks who will benefit from a 
+ * simple but fully-working example.
+ * My intent is to provide a complete, 
+ * self-contained introduction to modeling with Rcpp.  
+ * My hope is that this model can be easily modified 
+ * to run any dynamical simulation that has 
+ * dependence on the previous time step 
+ * (and can therefore not be vectorized).
+ *
+ * This post uses a classic 
+ * Susceptible-Infected-Recovered (SIR)
+ * [epidemiological compartment model](https://en.wikipedia.org/wiki/Compartmental_models_in_epidemiology).
+ * Compartment models are simple, commonly-used dynamical
+ * systems models.  Here I demonstrate the
+ * [tau-leap method](https://en.wikipedia.org/wiki/Tau-leaping), where
+ * a discrete number of individuals move probabilistically 
+ * between compartments at fixed intervals in time.
+ * In this model, the wait-times within class 
+ * are exponentially distributed, and the number of
+ * transitions between states in a fixed time step are 
+ * Poisson distributed.
+ *
+ * This model is parameterized for the spread of measles
+ * in a closed population, 
+ * where the birth rate (nu) = death rate (mu).
+ * The transmission rate (beta) describes how frequently 
+ * susceptible (S) and infected (I) individuals come into 
+ * contact, and the recovery rate (gamma) describes the
+ * the average time an individual spends infected before
+ * recovering. 
+*/
+
+/** 
+ * ### C++ Code
+ * Note that cpp Functions must be marked with the following comment like for use in R
+ * // [[Rcpp::export]]
+*/
+
+#include <Rcpp.h>
+// Declare commonly used objects
+using Rcpp::NumericVector;
+using Rcpp::List;
+using Rcpp::DataFrame;
+using Rcpp::Named;
+// use Rcpp namespace R:: (alias to R API)
+// R::rpois always returns a single value
+// to return multiple (e.g. Integer/NumericVector, 
+// use Rcpp::rpois(int ndraw, param) and friends
+using R::rpois;
+// rcpp also has a pmin for vectors
+using std::min;
+
+// This function will be used in R!
+// Evaluates the number of events 
+// and updates the states at each time step tau
+// [[Rcpp::export]]
+List tauleapCpp(List params){
+    // chained operations are tricky in cpp
+    // pull out list w/in list into its own object
+    List init = params["init"];
+    // need to "cast" R vector to cpp scalar
+    int nsteps = Rcpp::as<int>(params["nsteps"]);
+
+    // initialize each state vector in its own vector
+    // set all vals to initial vals
+    NumericVector SS(nsteps, init["S"]);
+    NumericVector II(nsteps, init["I"]);
+    NumericVector RR(nsteps, init["R"]);
+    NumericVector NN(nsteps, init["pop"]);
+    // fill time w/zeros
+    NumericVector time(nsteps);
+
+    // pull out params for easy reading 
+    double nu = params["nu"];
+    double mu = params["mu"];
+    double beta = params["beta"];
+    double gamma = params["gamma"];
+    double tau = params["tau"];
+
+    // Rcpp magic: Initialize RNG (uses R engine)
+    Rcpp::RNGScope scope;
```

Comment:
This is actually incorrect (it's done automatically by the attributes wrapper so is unnecessary). Realize that this exists in other examples and gallery articles and we are working to clean that up. 

Besides being redundant with the usage in the attributes wrapper, if used outside of an attributes wrapper it could cause a crash by triggering a GC after the destructor for the `sim` variable has run (which leaves the SEXP underneath `sim` unprotected for a GC that might be triggered by PutRNGState). 

Don't worry if this makes no sense, we'll have an article soon that explains all. In the meantime let's just remove the RNGScope here.